# Spam Detector

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [2]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

# My Prediction
## Random Forest Classifier
* Considering spam classification datasets often contain complex patterns and interactions between features, I believe a Random Forest Classifier is likely to perform better than Logistic Regression because Random Forest can handle non-linearities and feature interactions more effectively 

## Split the Data into Training and Testing Sets

In [7]:
# Create the labels set `y` and features DataFrame `X`
y = data['spam']
X = data.drop(columns=['spam'])
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

In [5]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
label_counts = y.value_counts()
print(label_counts)

spam
0    2788
1    1813
Name: count, dtype: int64


In [6]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1)
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

In [8]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1)
print(X_train.head())
print(y_train.head())

      word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
958             0.00               0.00           0.00           0.0   
1533            0.00               0.00           0.00           0.0   
654             0.33               0.00           0.33           0.0   
1497            0.52               1.31           0.26           0.0   
3844            0.00               0.00           0.00           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
958             0.0            0.00              1.21                0.00   
1533            0.0            0.00              0.00                0.00   
654             0.0            0.00              0.66                0.00   
1497            2.9            0.26              0.79                0.26   
3844            0.0            0.00              0.00                0.00   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
958               0.0  

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [9]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
scaler = StandardScaler()

In [14]:
# Fit the Standard Scaler with the training data
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [12]:
# Scale the training data
X_train_scaled_df = pd.DataFrame(X_train_scaled, columns=X_train.columns)
print(X_train_scaled_df.head())

   word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0       -0.335858          -0.160385      -0.550300     -0.046758   
1       -0.335858          -0.160385      -0.550300     -0.046758   
2        0.739868          -0.160385       0.092090     -0.046758   
3        1.359225           0.895761      -0.044174     -0.046758   
4       -0.335858          -0.160385      -0.550300     -0.046758   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0      -0.448094       -0.344231          2.950691           -0.252313   
1      -0.448094       -0.344231         -0.296875           -0.252313   
2      -0.448094       -0.344231          1.474525           -0.252313   
3       3.824612        0.636976          1.823437            0.377023   
4      -0.448094       -0.344231         -0.296875           -0.252313   

   word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
0        -0.326802        1.526950  ...             -0.115629 

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [13]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression


In [25]:
# Make and save testing predictions with the saved logistic regression model using the test data
model = LogisticRegression()
model.fit(X_train_scaled, y_train)
y_predict = model.predict(X_test_scaled)
prediction_df = pd.DataFrame({'ID': X_test.index,'Predicted_Label': y_predict})
prediction_df.to_csv("logistic_regression_predictions.csv", index=False)
# Review the predictions
saved_predictions = pd.read_csv("logistic_regression_predictions.csv")
print("LR Predictions Preview:")
print(saved_predictions.head())

LR Predictions Preview:
     ID  Predicted_Label
0  3683                0
1  4412                0
2  2584                0
3    69                1
4  1844                0


In [26]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy = accuracy_score(y_test, y_predict)
print(f"Logistic Regression Accuracy Score: {accuracy:.4f}")

Logistic Regression Accuracy Score: 0.9225


## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [27]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier


In [28]:
# Make and save testing predictions with the saved logistic regression model using the test data
rf_model = RandomForestClassifier(random_state=42)
rf_model.fit(X_train_scaled, y_train)
y_predict_rf = rf_model.predict(X_test_scaled)
predictions_rf_df = pd.DataFrame({'ID': X_test.index,'Predicted_Label': y_predict_rf})
predictions_rf_df.to_csv("random_forest_predictions.csv", index=False)
# Review the predictions
saved_rf_predictions = pd.read_csv("random_forest_predictions.csv")
print("RF Predictions Preview:")
print(saved_rf_predictions.head())

RF Predictions Preview:
     ID  Predicted_Label
0  3683                0
1  4412                0
2  2584                0
3    69                1
4  1844                0


In [29]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy_rf = accuracy_score(y_test, y_predict_rf)
print(f"Random Forest Classifier Accuracy Score: {accuracy_rf:.4f}")

Random Forest Classifier Accuracy Score: 0.9566


## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

# Model Evaluation
## Model Performance:
### Logistic Regression Model Accuracy:

Accuracy Score:  0.9225

### Random Forest Classifier Model Accuracy:

Accuracy Score:  0.9566

* My predciton was correct, Random Forest Classifier Model achieved a higher accuracy score. Random Forests Classifier can handle complex interactions and non-linear relationships between features better than Logistic Regression.